rspace/src/main/scala/coop/rchain/rspace/SpaceMatcher.scala

```diff
@@ -43,33 +44,32 @@ private[rspace] trait SpaceMatcher[F[_], C, P, E, A, R, K] extends ISpace[F, C,
     * along with the remaining unmatched data. If an illegal state is reached
     * during searching for a match we short circuit and return the state.
     */
-  @tailrec
   private[rspace] final def findMatchingDataCandidate(
       channel: C,
       data: Seq[(Datum[A], Int)],
       pattern: P,
       prefix: Seq[(Datum[A], Int)]
-  )(implicit m: Match[P, E, A, R]): Either[E, Option[(DataCandidate[C, R], Seq[(Datum[A], Int)])]] =
+  )(
+      implicit m: Match[F, P, E, A, R]
+  ): F[Either[E, Option[(DataCandidate[C, R], Seq[(Datum[A], Int)])]]] =
     data match {
       case (indexedDatum @ (Datum(matchCandidate, persist, produceRef), dataIndex)) +: remaining =>
-        m.get(pattern, matchCandidate) match {
+        m.get(pattern, matchCandidate).flatMap {
           case Left(ex) =>
-            Left(ex)
+            ex.asLeft[Option[(DataCandidate[C, R], Seq[(Datum[A], Int)])]].pure[F]
```

Comment:
There's no need to unwrap `ex` and wrap it back into a `Left`. `case result: Left => result.pure[F]` should suffice

Reply:
There's no need to have `Left` at all 👍